qip-0009.md

```diff
@@ -0,0 +1,529 @@
+3:    return VerifyMerkleRoot(proof.proofHeaders[0], proof.merkleProof, proof.txHash)
+4: end function
+
+```
+```
+// Calculates the total intrinsic work of a array of headers and return it
+1: function TotalIntrinsicOfProof(proof)
+2:    total ← 0  // Initialize the total intrinsic value
+
+3:    for each header in proof.proofHeaders do
+4:        // Add the intrinsic work of each header to the total
+5:        total += header.IntrinsicWork()
+6:    end for
+
+7:    return total  // Return the total intrinsic value
+8: end function
+
+```
+
+
+## Parameters and Implications
+
+Since there is non-negligible cost to the size of interlink structure and number of levels to be stored in the body of a block, its good to constrain its size. It is also important to look at the 
+Work compression and speed/cost of producing proofs. Given these constraints and considering the fact that Quai will solve the problem of scaling work based blockchains and becomes the global monetary system that runs for atleast 30 more years, I think we want to pick $k=4$ and $l=4$. This means we would have 4 levels of interlinks and interlinks would have $(target + (4, 8, 12, 16)$ extra zeroes). On expectation the age of the interlinks would be every $interlink[0]=16$ prime blocks, $interlink[1]=256$ prime blocks, $interlink[2]=4096$ prime blocks, and $interlink[3]=65,536$ prime blocks.
+
+## Proof size
+If r is the number of Region blocks and z is the number of zone blocks per region and timeFactor(t) is the ratio that modulates the frequency of occurance of Dom Blocks.
+* Expected number of Zone blocks per Region block - $z \times t$
+* Expected number of Region blocks per Prime block - $r \times t$
+* Frequency of a Prime blocks from a particular zone - Every $ r\times z$ prime blocks
+* Frequency of interlinks based on the rank - $2^{(k \times \mu)}$
+
+Lets say a proof needs to be produced for a tx inclusion j blocks ago, and current height is n, and m is the security parameter. The expected size of the proof is 
+
+Let p the number of prime blocks
+
+$$ p = \frac{n-j}{r \times z \times t^2}$$
+
+Now the number of blocks for a given rank of interlink can be calculated
+
+$$\frac{p}{2^{k * \mu}}$$
+
+This means the number of interlink blocks in the proof will be 
+
+$$\text{interlinksize} = \sum_{i=0}^{i=l} \left\lfloor  \frac{\frac{p}{2^{k*i}}}{m} \right\rfloor $$
+
+If on expectation there can be $z \times t$ zone between finding a Region blocks, and $r \times t$ number of Region Blocks, worst case for a long proof will be
+
+$$\text{proofsize} = 2zt + 2rt +  \sum_{i=0}^{i=l} \left\lfloor  \frac{p}{2^{k*i} \times m} \right\rfloor$$
+ number of blocks. The factor of 2 comes because tracing up from both sides to the Prime block.
+
+## Future work
+Snap Sync using interlinks will be proposed in a future QIP.
+
+## References
+
+[1]: Non-Interactive Proofs of Proof-of-Work  https://eprint.iacr.org/2017/963.pdf
```

Comment:
Might also add a reference to the initial interlink idea:
https://fc16.ifca.ai/bitcoin/papers/KLS16.pdf

Reply:
okay

Reply:
added it